drivers/ethernet/eth_e1000.c

```diff
@@ -0,0 +1,340 @@
+	RAH	= 0x5404,	/* Receive Address High */
+};
+
+/* Legacy TX Descriptor */
+struct e1000_tx {
+	u64_t addr;
+	u16_t len;
+	u8_t  cso;
+	u8_t  cmd;
+	u8_t  sta;
+	u8_t  css;
+	u16_t special;
+};
+
+/* Legacy RX Descriptor */
+struct e1000_rx {
+	u64_t addr;
+	u16_t len;
+	u16_t csum;
+	u8_t  sta;
+	u8_t  err;
+	u16_t special;
+};
+
+struct e1000_dev {
+	volatile struct e1000_tx tx __aligned(16);
+	volatile struct e1000_rx rx __aligned(16);
+	struct pci_dev_info pci;
+	struct net_if *iface;
+	u8_t mac[ETH_ALEN];
+	u8_t txb[E1000_MTU];
+	u8_t rxb[E1000_MTU];
+};
+
+static const char *e1000_reg_to_string(enum e1000_reg_t r);
+static void e1000_init(struct net_if *iface);
+
+#define W(_dev, _reg, _val) do {				\
+	D("W %s 0x%08x", e1000_reg_to_string(_reg), _val); 	\
+	sys_write32(_val, (_dev)->pci.addr + _reg);		\
+} while (0)
+
+#define R(_dev, _reg)						\
+({								\
+	u32_t val = sys_read32((_dev)->pci.addr + (_reg));	\
+	D("R %s 0x%08x", e1000_reg_to_string(_reg), val); 	\
+	val;							\
+})
+
+static const char *e1000_reg_to_string(enum e1000_reg_t r)
+{
+#define _(_x)	case _x: return #_x
+	switch (r) {
+	_(CTRL);
+	_(ICR);
+	_(ICS);
+	_(IMS);
+	_(RCTL);
+	_(TCTL);
+	_(RDBAL);
+	_(RDBAH);
+	_(RDLEN);
+	_(RDH);
+	_(RDT);
+	_(TDBAL);
+	_(TDBAH);
+	_(TDLEN);
+	_(TDH);
+	_(TDT);
+	_(RAL);
+	_(RAH);
+	}
+	E("Unsupported register: 0x%x", r);
+	k_oops();
+	return NULL;
+}
+
+static enum ethernet_hw_caps e1000_caps(struct device *dev)
+{
+	return  ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T | \
+		ETHERNET_LINK_1000BASE_T;
+}
+
+static size_t e1000_linearize(struct net_pkt *pkt, void *buf, size_t bufsize)
+{
+	size_t len = net_pkt_ll_reserve(pkt) + pkt->frags->len;
+	struct net_buf *nb;
+
+	/* First fragment contains link layer (Ethernet) header */
+	memcpy(buf, net_pkt_ll(pkt), len);
+
+	for (nb = pkt->frags->frags; nb; nb = nb->frags) {
+		memcpy(buf + len, nb->data, nb->len);
+		len += nb->len;
+	}
+
+	return len;
+}
+
+static int e1000_tx(struct e1000_dev *dev, void *data, size_t data_len)
+{
+	dev->tx.addr = POINTER_TO_INT(data);
+	dev->tx.len = data_len;
+	dev->tx.cmd = TDESC_EOP | TDESC_RS;
+
+	W(dev, TDT, 1);
+
+	while(!(dev->tx.sta)) {
+		k_yield();
+	}
+
+	D("tx.sta: 0x%02hx", dev->tx.sta);
+
+	return (dev->tx.sta & TDESC_STA_DD) ? 0 : -EIO;
+}
+
+static int e1000_send(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct e1000_dev *dev = net_if_get_device(iface)->driver_data;
+
+	size_t len = e1000_linearize(pkt, dev->txb, sizeof(dev->txb));
+
+	net_pkt_unref(pkt);
+
+	return e1000_tx(dev, dev->txb, len);
+}
+
+static struct net_pkt *e1000_rx(struct e1000_dev *dev)
+{
+	struct net_pkt *pkt = NULL;
+
+	D("rx.sta: 0x%02hx", dev->rx.sta);
+
+	if (!(dev->rx.sta & RDESC_STA_DD)) {
+		E("RX descriptor not ready");
+		goto out;
+	}
+
+	pkt = net_pkt_get_reserve_rx(0, K_NO_WAIT);
+	if (!pkt) {
+		E("Out of RX buffers");
+		goto out;
+	}
+
+	union {
+		u64_t qw;
+		u32_t dw[2];
+	} addr = { .qw = dev->rx.addr };
```

Comment:
We don't allow declaring a variable in the middle of a code block, so move that above